File: starter_kits/Swift/Halite-III/Halite-III/hlt/Direction.swift
//
//  Direction.swift
//  Halite-III
//
//  Created by Chris Downie on 10/24/18.
//  Copyright © 2018 Chris Downie. All rights reserved.
//

import Foundation

enum Direction: CaseIterable {
    case north, west, south, east
    case still
    
    /// Returns an array of all cardinal directions
    ///
    /// - Returns: An array of all cardinal directions
    static func getAllCardinals() -> [Direction] {
        return [.north, .west, .south, .east]
    }
    
    /// Returns a letter command from a direction
    ///
    /// - Parameter direction: The direction to convert to a
    /// - Returns: A command equivalent to the given direction
    static func convert(_ direction: Direction, forShipId shipId: Ship.ID) -> Command {
        return .move(shipId: shipId, direction: direction)
    }
    
    /// Returns the opposite cardinal direction given a direction
    ///
    /// - Parameter direction: The input direction
    /// - Returns: The opposite direction
    static func invert(_ direction: Direction) -> Direction {
        switch direction {
        case .north:
            return .south
        case .south:
            return .north
        case .east:
            return .west
        case .west:
            return .east
        case .still:
            return .still
        }
    }
}

// MARK: - Convenience accessors
// This doesn't match the letter of the API, but it feels like a more Swift-friendly means of providing the same
// functionality.
extension Direction {
    static prefix func -(_ direction: Direction) -> Direction {
        return Direction.invert(direction)
    }
    
    static var allCardinals = {
        return Direction.getAllCardinals()
    }()
}

File: starter_kits/Clojure/hlt/lib/hlt/direction.clj
(ns hlt.direction
  (:require [hlt.commands :as commands]))

(def north [0 -1])
(def south [0 1])
(def east [1 0])
(def west [-1 0])

(def still [0 0])

(def all-cardinals [north south east west])

(defn get-all-cardinals
  []
  all-cardinals)

(defn convert
  "Converts from this direction tuple notation to the engine's string notation"
  [direction]
  (cond
    (= direction north) commands/north
    (= direction south) commands/south
    (= direction east) commands/east
    (= direction west) commands/west
    :else
    commands/stay-still))

(defn invert
  "Returns the opposite cardinal direction given a direction"
  [direction]
  (cond
    (= direction north) south
    (= direction south) north
    (= direction east) west
    (= direction west) east
    :else
    still))
